services/openvpn/service/manager.go

```diff
@@ -0,0 +1,146 @@
+	"github.com/mysteriumnetwork/node/core/ip"
+	"github.com/mysteriumnetwork/node/core/location"
+	"github.com/mysteriumnetwork/node/identity"
+	"github.com/mysteriumnetwork/node/nat"
+	dto_discovery "github.com/mysteriumnetwork/node/service_discovery/dto"
+	"github.com/mysteriumnetwork/node/session"
+)
+
+const logPrefix = "[service-openvpn] "
+
+// ServerFactory initiates Openvon server instance during runtime
+type ServerFactory func(primitives *tls.Primitives) openvpn.Process
+
+// ProposalFactory prepares service proposal during runtime
+type ProposalFactory func(currentLocation dto_discovery.Location) dto_discovery.ServiceProposal
+
+// SessionManagerFactory initiates session manager instance during runtime
+type SessionManagerFactory func(primitives *tls.Primitives, outboundIP, publicIP string) session.Manager
+
+// Manager represents entrypoint for Openvpn service with top level components
+type Manager struct {
+	ipResolver            ip.Resolver
+	natService            nat.NATService
+	locationResolver      location.Resolver
+	proposalFactory       ProposalFactory
+	sessionManagerFactory SessionManagerFactory
+
+	vpnServerFactory ServerFactory
+	vpnServer        openvpn.Process
+}
+
+// Start starts service - does not block
+func (manager *Manager) Start(providerID identity.Identity) (
+	proposal dto_discovery.ServiceProposal,
+	sessionManager session.Manager,
+	err error,
+) {
+	publicIP, err := manager.ipResolver.GetPublicIP()
+	if err != nil {
+		return
+	}
+
+	// if for some reason we will need truly external IP, use GetPublicIP()
+	outboundIP, err := manager.ipResolver.GetOutboundIP()
+	if err != nil {
+		return
+	}
+
+	manager.natService.Add(nat.RuleForwarding{
+		SourceAddress: "10.8.0.0/24",
+		TargetIP:      outboundIP,
+	})
+
+	err = manager.natService.Start()
+	if err != nil {
+		log.Warn(logPrefix, "received nat service error: ", err, " trying to proceed.")
+	}
+
+	currentCountry, err := manager.locationResolver.ResolveCountry(publicIP)
+	if err != nil {
+		return
+	}
+	currentLocation := dto_discovery.Location{Country: currentCountry}
+	log.Info(logPrefix, "Country detected: ", currentCountry)
+
+	caSubject := pkix.Name{
+		Country:            []string{currentCountry},
+		Organization:       []string{"Mysterium Network"},
+		OrganizationalUnit: []string{"Mysterium Team"},
+	}
+	serverCertSubject := pkix.Name{
+		Country:            []string{currentCountry},
+		Organization:       []string{"Mysterium node operator company"},
+		OrganizationalUnit: []string{"Node operator team"},
+		CommonName:         providerID.Address,
+	}
+
+	primitives, err := tls.NewTLSPrimitives(caSubject, serverCertSubject)
+	if err != nil {
+		return
+	}
+
+	manager.vpnServer = manager.vpnServerFactory(primitives)
+	if err = manager.vpnServer.Start(); err != nil {
+		return
+	}
+
+	proposal = manager.proposalFactory(currentLocation)
+	sessionManager = manager.sessionManagerFactory(primitives, outboundIP, publicIP)
+	err = nil
```

Comment:
this prob is set to nil at initialisation. Excessive.

Reply:
Fixed